File: 2025_go/07_laboratories.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	fmt.Println("Sample A", solve07a("inputs/07_sample.txt", false)) // 21
	fmt.Println("A", solve07a("inputs/07_input.txt", false))         // 1678
	fmt.Println("Sample B", solve07a("inputs/07_sample.txt", true))  // 40
	fmt.Println("B", solve07a("inputs/07_input.txt", true))          // 357525737893560
}

func solve07a(filename string, b bool) int {
	var (
		f, _    = os.Open(filename)
		scanner = bufio.NewScanner(f)
		splits  = 0
		count   []int
	)
	defer f.Close()
	for scanner.Scan() {
		line := scanner.Text()
		if count == nil {
			count = make([]int, len(line))
		}
		for x := range line {
			switch line[x] {
			case 'S':
				count[x] = 1
			case '^':
				count[x-1] += count[x]
				count[x+1] += count[x]
				if count[x] > 0 {
					splits++
				}
				count[x] = 0
			}
		}
	}
	if b {
		paths := 0
		for x := range count {
			paths += count[x]
		}
		return paths
	}
	return splits
}

File: 2025_go/07_laboratories.py
#!/usr/bin/env python3


def solve07(lines: list[str], b: bool) -> int:
    count = [0] * len(lines[0])
    splits = 0

    for line in lines:
        for x, c in enumerate(line):
            if c == 'S':
                count[x] = 1
            elif c == '^':
                count[x-1] += count[x]
                count[x+1] += count[x]
                splits += count[x] > 0
                count[x] = 0

    return sum(count) if b else splits


if __name__ == "__main__":
    with open("inputs/07_sample.txt") as f:
        s = [l for line in f if (l := line.strip())]
    with open("inputs/07_input.txt") as f:
        i = [l for line in f if (l := line.strip())]

    print(f"Sample A: {solve07(s, False)}")  # 21
    print(f"A:        {solve07(i, False)}")  # 1678
    print(f"Sample B: {solve07(s, True)}")  # 40
    print(f"B:        {solve07(i, True)}")  # 357525737893560
